### ISSUE TITLE
Expand the ' broadcast' function to combine variables with 1+ dimensions

### ISSUE THREAD

**Issue by adrivinca** [iiasa/message-ix-models#73]
Hi, @awais307 and I often defined new parameters for new technologies using the current ` broadcast`  function, which basically expands an existing dataframe for an additional dimension.
However, we often use it to expand the year_vtg and year_act dimensions, which are also linked to a mapping that takes lifetime into account (see [issue #571 in message_ix](https://github.com/iiasa/message_ix/issues/571) )

if we use `broadcast `for the sets ` year_vtg`  and ` year_act`  separately, we define also a number of vtg-act years that are not needed

```python
import pandas as pd
from message_ix import make_df
from message_ix_models.util import broadcast, private_data_path
    
inp0 = make_df(
        "input",
            technology="irrigation_oilcrops",
            value=1,
            unit="-",
            level="water_supply",
            commodity="freshwater",
            mode="M1",
            time="year",
            time_origin="year",
            node_origin=['AFR','NAM'],
    )
inp1 = inp0.pipe(broadcast, year_vtg=[2010,2020], 
           year_act=[2010,2020])
```
in this case, inp1 is wrong

what we would like to achieve is basically the combination of the dataframe with the missing variables `inp0` and the mapping dataframe between `year_vgt` and ` year_act`

```python
map_v_a = pd.DataFrame({
    'year_vtg' : [2010,2010,2020],
    'year_act' : [2010,2020,2020] }).astype(object)

inp2 = broadcast(inp0,toadd = map_v_a)
```

of course it would not work with the current version of `broadcast`, but I edited it so that it could work (see below).

I wonder if this is the way to go for you, if we should make a dedicated function, or if anybody have suggestions to make the code 'cleaner'. 

Notes:

- I explored the `merge` option vs `concat` and `merge` seems simpler with multiple variables
- the `print()` are just for testing it

```python
def broadcast(df,toadd = None, **kwargs):
    """Fill missing data in `df` by broadcasting.

    Arguments
    ---------
    kwargs
        Keys are dimensions. Values are labels along that dimension to fill.
    """
    print('KWARGS: ',kwargs)
    if toadd is not None:
        print('dataframe')

        for cc in toadd.columns:
            if cc in df.columns:
                assert df[cc].isna().all(), f"Dimension {dim} was not empty\n\n{df.head()}"
                df.drop(cc, axis=1,inplace = True)
        
        #concat_data = pd.concat([df,to_add['year_vtg']],axis = 1)
        df['tmp'] = 1
        toadd['tmp'] = 1
        df = pd.merge(df,toadd).drop(columns ={'tmp'})

    else:
        print('single parameter')
    
        for dim, levels in kwargs.items():
            # Checks
            assert df[dim].isna().all(), f"Dimension {dim} was not empty\n\n{df.head()}"
            if len(levels) == 0:
                log.debug(
                    f"Don't broadcast over {repr(dim)}; labels {levels} have length 0"
                )
                continue
    
            # - Duplicate the data
            # - Drop the existing column named 'dim'
            # - Re-add the column from the constructed MultiIndex
            # - Reindex for sequential row numbers
            df = (
                pd.concat([df] * len(levels), keys=levels, names=[dim])
                .drop(dim, axis=1)
                .reset_index(dim)
                .reset_index(drop=True)
            )
    return df
```

**Comment by khaeru** [comment]
This is a good idea, thanks!

**Comment by khaeru** [comment]
I think we can take advantage of the way `.pipe()` works: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.pipe.html?highlight=pipe#pandas.DataFrame.pipe —note that it accepts **either** keyword arguments *or* positional arguments to be passed to the called `func`.

Currently `broadcast()` uses the keyword arguments:
https://github.com/iiasa/message-ix-models/blob/6929fd6df95388dc3635a56642688f9f00155d1b/message_ix_models/util/__init__.py#L164
….and understands that the argument name == the dimension/column name to be filled in the df.

I think the implementation could be very similar to what you suggest:
- Instead of a keyword argument `toadd`, make this an optional, single, positional argument, maybe `labels: pd.DataFrame` (labels for multiple dimensions).
- Instead of an `if:` block, allow both to be used simultaneously: first `labels` is handled, and then the existing code for any 1 or more dimensions in`kwargs`.

**Comment by khaeru** [comment]
@adrivinca can we consider this closed by #74?

**Comment by adrivinca** [comment]
yes, thanks for pointing it out
